### ISSUE TITLE
Permission to use your patches in SymPy

### ISSUE THREAD

**Issue by certik** [diofant/diofant#432]
Hi Sergey,

Due to popular demand, people are asking if SymPy can merge some of your patches back to SymPy. I have created a PR (https://github.com/sympy/sympy/pull/10531) which copies your BSD license into SymPy, and starts using your patches. The arrangement would be that in the future, we can just merge patches from your "master" branch, after that PR is merged.

I would like to get your explicit approval that we can do that. Are you ok with this arrangement?

Thanks,
Ondrej

**Comment by skirpichev** [comment]
@certik, I don't believe that I should give explicit permissions to every person and/or project, which will take code from the Diofant project.

Have you read the project's [LICENSE](https://github.com/diofant/diofant/blob/master/LICENSE) file?  So far, if 1) there is nothing wrong with this file and 2) you understand terms of the LICENSE - I suggest you just obey all LICENSE terms.

> sympy/sympy#10531

From my understanding of the BSD conditions (I'm not a lawyer) - it's that you need.  Keep in mind, that you
must synchronize copied LICENSE with Diofant's one, from time to time.  Besides that, probably you should mention in commit messages/PR's, that you are taking piece of other people code...

Please reopen issue, if you think there is still something I should clarify.  Or if you think - there is something wrong with this file.  E.g. with copyright statements.

**Comment by skirpichev** [comment]
BTW, as it's expected, all this was just another farce from sympy owners...  Most likely, now they close  sympy/sympy#10531 with "independent solution" (sympy/sympy#12182) and then continue spread the lie that "the author has not given us permission to merge" (c)

**Comment by certik** [comment]
Thank you @skirpichev for the quick answer, I think that fully answers my question, right now I don't have any other questions --- if something comes up, I will let you know by commenting here or opening a new issue.

**Comment by skirpichev** [comment]
@certik, given that you already started edit the project history, by removing "wrong persons" (c.f. [this](https://github.com/sympy/sympy/graphs/contributors) and [page in my project](https://github.com/diofant/diofant/graphs/contributors) - I doubt that you were interested in any answers.  Except, maybe, a direct permission for using code from this project without any acknowledgement (as you already do now).

I'll be very surprised if you will not just continue this practice.  Mentioned pr by jksuom - just another proof of that, I don't believe in such "accidents".  I hope, this will kick your ass someday...

> I will let you know by commenting here

If you are expecting some actions from me (e.g. LICENSE fixing) - please reopen this issue (or open a new one), not just add comments.

**Comment by certik** [comment]
@skirpichev I can't remove any people from the history on github, it's a bug in their system. I already contacted GitHub twice if they could fix it, but they didn't get back to me, so I created an issue for this: https://github.com/sympy/sympy/issues/12198.

I assume you accuse me of contacting github to remove you from history. I haven't done anything like that. In fact, I invited you to be a co-author on our common article (https://peerj.com/articles/cs-103/), and the order of authors is sorted by the number of commits, so you are the 5th author there. You accusation is not based on facts.

Regarding jksuom, I haven't talked with him at all about any PRs regarding this issue. Again, your accusation is false.

I am trying to resolve it. I was waiting for your answer for over a year. Now when you answered, the ball is in our court, I am talking with others what the next steps should be. I trust that a few days for us to figure out the next steps is not a problem for you, given the one year delay on your part.

**Comment by skirpichev** [comment]
> I haven't done anything like that.

Maybe Aaron.  I can't guess who did this, but I have reply from github (very quick in fact), that it's not a bug in their system and is "a result of configurations or settings performed by the repository owner", which they will not explain ("We do consider these types of actions confidential, and will not be able to share any details beyond this.").

> You accusation is not based on facts.

My [answer above](https://github.com/diofant/diofant/issues/432#issuecomment-280827713) lists all facts.

> Regarding jksuom, I haven't talked with him at all about any PRs regarding this issue.

Please don't take this personally.  Of course, I can't tell who exactly from SymPy
owners asked him.  But the whole story is a fact and it's not looking to be an accident.  Lets see if my prediction will be correct (you will merge new "independent" solution and close old request).

> I was waiting for your answer for over a year.

You had same answer as [above](https://github.com/diofant/diofant/issues/432#issuecomment-279731483) - year ago, already before your private questions.  For example, [this](https://groups.google.com/d/msg/sympy/pfYJI-lxkYE/tDx5d-v0AwAJ).  (All that is still in public, for now...)  And I can easily remind you a letter with a sentence like "I'm not planning to discuss anything related to my open-source projects privately."

> I am talking with others

Privately, as usual.  That's exactly why mentioned above pr by jksuom - doesn't seems to be an accident for me...

**Comment by skirpichev** [comment]
@certik, please update sympy/sympy#10531 to confirm that you got answer.

**Comment by skirpichev** [comment]
@cbm755, as you apparently interested in merging some code, same request for you: could you update sympy/sympy#10531 to confirm that sympy devs got an answer?

**Comment by certik** [comment]
@skirpichev yes, I was discussing the situation with other people, and I just updated it:

https://github.com/sympy/sympy/pull/10531#issuecomment-281401189

I think the situation is clear, except a minor thing (#445), but that doesn't prevent us from merging https://github.com/sympy/sympy/pull/10531. The reason I raised #445 is if you decide to change the license, so that I can modify the license in https://github.com/sympy/sympy/pull/10531 before merging. Let me know.

Thank you @skirpichev for answering the question here, I think all is clear, and I would like to start using your improvements back in SymPy, rather than providing our own fixes.

**Comment by cbm755** [comment]
> confirm that sympy devs got an answer?

@skirpichev speaking only for myself: I think you answered a year ago, stating anyone can use your code provided they respect the license.  This is self-evident, and indeed no explicit permission is needed.

So I've tried to interpret subsequent discussion to mean "how exactly is that done; what are the mechanics of attribution and respecting the license".

But beyond the strict "letter of the license" (which again, obviously must be satisfied), I would personally be happier if you felt your work was being appropriately attributed.  That is, a matter of politeness rather than law.  I try to interpret other people's actions/statements by assuming they feel the same.

**Comment by skirpichev** [comment]
On Tue, Feb 21, 2017 at 12:37:41PM -0800, Colin B. Macdonald wrote:
>      confirm that sympy devs got an answer?
> 
>    [1]@skirpichev speaking only for myself: I think you answered a year ago,
>    stating anyone can use your code provided they respect the license. This
>    is self-evident, and indeed no explicit permission is needed.
> 
>    So I've tried to interpret subsequent discussion to mean "how exactly is
>    that done; what are the mechanics of attribution and respecting the
>    license".
> 
>    But beyond the strict "letter of the license" (which again, obviously must
>    be satisfied), I would personally be happier if you felt your work was
>    being appropriately attributed. That is, a matter of politeness rather
>    than law. I try to interpret other people's actions/statements by assuming
>    they feel the same.

Sorry for disturbing you.  It seems, sympy/sympy#10531 was updated
by Ondrej today (by accident, after my request to you).

**Comment by certik** [comment]
Yes, what @cbm755 said is correct. There is no ill will, at least not on my side, but I think there is no ill will on other sympy developers either, as @cbm755 said. I just want to make sure that all parties are on board with the solution, and I think we reached an agreement.

P.S. @skirpichev I didn't forget about that contributors page issue, I urged GitHub again today:

https://github.com/sympy/sympy/issues/12198#issuecomment-281706226

Feel free to email them too and reference that issue. I have no idea why they do not get back to me.

**Comment by skirpichev** [comment]
> There is no ill will

Just another accident.  Maybe.

> Feel free to email them too and reference that issue.

I asked them before.  They told me: that this was "a result of configurations or settings performed by the repository owner" and "we do consider these types of actions confidential, and will not be able to share any details beyond this."

By encouraging of such things GitHub looses my respect at some rate.  But on another hand, it's a useful service, that works (for me) for free...  And I don't think that I can distract their support team by repeating clearly answered questions.

**Comment by certik** [comment]
But what "configuration or setting" does that? The only setting that we did was to block the user skirpichev, and that should *not* delete such a user from history. But just in case, I unblocked the user skirpichev (https://github.com/sympy/sympy/issues/12198#issuecomment-281725807), and you are still missing from the contributors page. So if you know about any further option that I should change at this point, please let me know.

**Comment by skirpichev** [comment]
> So if you know about any further option that I should change at this point, please let me know.

No, I don't know.  I tried blocking for few mins - but that doesn't affect statistics.

**Comment by certik** [comment]
https://github.com/sympy/sympy/pull/10531 is merged. Thanks @skirpichev for the feedback. Hopefully this solution will lead to everybody's benefit.

**Comment by skirpichev** [comment]
I doubt.  Codebase was diverged already too much, projects have different goals, merging policy, etc.  But lets see.

FYI, I opened mpmath's issue about eliminate=True defaults for hyper().

**Comment by skirpichev** [comment]
Wow, it seems you ban me again.  I tried posting to sympy/sympy#14210 and sympy/sympy#14738 - comments just disappear.  I didn't know such is possible.

@certik, I wonder, why are you doing this?  Did some my recent comment violate your CoC?  Which one?

**Comment by skirpichev** [comment]
@cbm755, if you still are sympy contributor - could I ask you again to be my voice in sympy community in this case too?  Unfortunately, it's very possible that Ondrej just ignore my questions.  Of course, if you support silent blocking of contributors - sorry for asking you...

I don't post prs against sympy repo, but I fill some bugs, if they seems to be valid both for diofant and sympy, see e.g. [here](https://github.com/sympy/sympy/issues?q=is:issue+author:skirpichev+created:>2016-01-01+is:open).  (BTW, only one )  I believe, these bugs are more a benefit for sympy then for me (and diofant project).

**Comment by cbm755** [comment]
> if you support silent blocking of contributors

Of course I do not.

I have no idea what is happening in this instance, but my personal philosophy on all such matters is [never attribute to malice what can be explained by accident or incompetence](https://en.wikipedia.org/wiki/Hanlon%27s_razor).

**Comment by skirpichev** [comment]
Anyway, thank you for help, @cbm755.

**Comment by skirpichev** [comment]
> sigh I don't want to get involved in this...

@cbm755, please don't be.  Collin, I'm sorry to bother you.  I contacted you first, because previously you helped me in similar case.  My appeal was in hope that it could be useful for sympy community, but it seems that silent blocking is consistent with your expectations from the project.

Ok.  I will wait for while and then - try to ask other sympy developer.  Perhaps someone else find this situation wrong and contact superiors to make this case public.

> I have not "signed" anything.

But you do agree on https://github.com/sympy/sympy/pull/10506, there is no negative review from you.

> If your comment was a violation, it could be hidden-by-default or otherwise flagged as such (we've seen that on other threads).

Perhaps, it's too severe violation, if project admins so silent and can't even mark wrong comment as wrong.

> your comment would personally bother me but its Free Software so it does not.

From all above, probably, I can't ask what exactly may bother you in this comment.  I guess, it's "diofant doesn't have "Code of Conduct" to ban you from any critique - feel free to say anything you do think about the diofant project".  But your CoC is used here right for this "goal".  (BTW, I won't be surprised if @JohnStone2017 now also banned.)  I would like to agree that bugreport was very emotional, but it mention real problems.  Why just not ignore the rest?  At least, issue shouldn't be just closed unsolved.

> Can you still not post to the issues you mentioned earlier in #432 (comment)?

Given that you (as you are member of sympy devs) start polishing project history again (see [contributors](https://github.com/sympy/sympy/graphs/contributors) page) - no, I'm still blocked.

**Comment by cbm755** [comment]
>> your comment would personally bother me but its Free Software so it does not.

> From all above, probably, I can't ask what exactly may
> bother you in this comment. I guess, it's "diofant doesn't
> have "Code of Conduct" to ban you from any critique - feel
> free to say anything you do think about the diofant project". 

No, I have no problem with that.  I should have clarified: what I would find distasteful is the promotion of a competing proprietary software project "B" in the bug report of Free project "A".  However, in this instance, "B" is Free Software; so I feel we're all in same bigger umbrella of creating Free mathematics software together.

Analogy: if you went on an Emacs bug report and said "you should use Vim, it can do this better"; I'd probably just shrug.  Whereas if you said "MS Word can do this better" I'd be... well I'd be upset ;-)

**Comment by cbm755** [comment]
> silent blocking is consistent with your expectations from the project.

FTR, it is not.

> that bugreport was very emotional, but it mention real problems. Why just not ignore the rest? At least, issue shouldn't be just closed unsolved.

I think you are correct, but volunteers have no duty to pay attention to any particular bug.  I'm not paying attention to that one (not interested in that technical issue).  And if a stranger acts rude, I generally pay even less attention.

**Comment by skirpichev** [comment]
> I should have clarified: what I would find distasteful is the promotion of a competing proprietary software project "B" in the bug report of Free project "A".

Ok, got it.  But in fact, this "promotion" barely explains how the reported problem could be solved.  Well, at least how it can be detected while CI build.

> Analogy: if you went on an Emacs bug report and said "you should use Vim, it can do this better";

I doubt it's a correct analogy.  Rather, something like this: "Emacs has critical issues, which aren't going to be fixed, I'm totally disappointed your release practice.  I'm thinking to use Vim" with reply "Try XEmacs, it's a fork, which doesn't have such problems due to following technical reasons [...]".

Anyway, even if there is some promotion of competing open source software - how does it violate your CoC?

Ok, @cbm755, I assume you can't help.  

@smichr, I'm asking you.  Could you kindly contact your superiors (Ondrej or Aaron) to ask them to provide public (in maillist or appropriate thread of your bugtracker) explanation of reasons for my ban?  Which wrong comment I did, how it does violate your CoC?  I don't expect they unban me or stop wiping me from the project history (see [this](https://github.com/sympy/sympy/graphs/contributors)), but I believe such action as banning former contributor - shouldn't be silent.  Can't you be next?

**Comment by cbm755** [comment]
> how does it violate your CoC?

I did not mean to imply it violated anything, just that the comment was mildly unpleasant for me.

> ... start polishing project history again (see contributors page)

This "polishing history" bit was bothering me; so I went back read https://github.com/sympy/sympy/issues/12198 specifically https://github.com/sympy/sympy/issues/12198#issuecomment-282321446.  It would appear that disappearing from the contributors page is a **undesired** side effect of blocking someone (this is apparently Github's bug, not SymPy's).

Perhaps there are other such bugs but the most likely explanation is you are indeed blocked.  I make no comment either way about the appropriateness of such a block (I am biased because you were my mentor in SymPy).

**Comment by skirpichev** [comment]
Ok, @smichr, I assume you also can't help.

@Upabjojr, I'm asking you. Could you kindly contact your superiors (Ondrej or Aaron) to ask them to provide public (in maillist or appropriate thread of your bugtracker) explanation of reasons for my ban? Which wrong comment I did, how it does violate your CoC? I don't expect they unban me or stop wiping me from the project history (see this), but I believe such action as banning former contributor - shouldn't be silent. Can't you be next?

**Comment by Upabjojr** [comment]
I'm not aware of any ban. Which site are you unable to access?

**Comment by skirpichev** [comment]
@Upabjojr, sympy bugtracker.

**Comment by bjodah** [comment]
I was blocked by Github for a short while in 2017, I contacted them and "fixed" the issue (apparently some bot of theirs had flagged my account as "suspicious" based on god knows what).

**Comment by skirpichev** [comment]
> I was blocked by Github for a short while in 2017, I contacted them and they "fixed" the issue

@bjodah, this is not related to the specific project, I guess, as in my case?  So, I think my situation is different: your superiors just silently blocked me.  I asked Ondrej first to comment this problem, but apparently he is not willing to respond.

@Upabjojr, thank you.

**Comment by skirpichev** [comment]
Wow.  They finally ["explained"](https://github.com/sympy/sympy/issues/14925#issuecomment-407275207) things.  The [actual bug](https://github.com/sympy/sympy/issues/14924) in which was my comment - they removed totally (and, btw, also a comment from @JohnStone2017 in  https://github.com/sympy/sympy/issues/14925).

You have great leaders, guys - @certik and Co don't provide you even a chance to check validity of their motivations and feed community with a direct lie.

Also looks to be a brilliant way to fix bugs.  Given the numbers of dead bugreports in your repo - I think it's an excellent strategy to fix them...

**Comment by cbm755** [comment]
> they removed totally

Maybe, and perhaps other explanations are also possible.  But to keep an accurate technical archive, I have filed https://github.com/sympy/sympy/issues/14966 with my understanding of the missing technical discussion.

**Comment by skirpichev** [comment]
> Maybe, and perhaps other explanations are also possible.

@cbm755, are there others?  It's possible from the github interface to remove comment, but not the issue as whole.  Thus, it's easy to find who did this.

> sympy/sympy#14966 with my understanding of the missing technical discussion.

Oh, of course, it's possible to fix other part of the issue, which is due to naming conflict for series function vs series module.  It's not "an unfortunate side effect of the way Python names work".

Python works fine.  It's a bug.  Which **can be fixed**.  If instead of removing unwanted comments and then blaming for "trolling and derogatory comments" (sorry, but @certik "explanation" without any proofs was extremely cynical) - sympy developers start fixing old bugs.  E.g. sympy/sympy#5362.
